Language/Java/Transformer.hs

```diff
@@ -0,0 +1,391 @@
+
+addFormalParams :: IdentCollection -> [Ident] -> IdentCollection
+addFormalParams identCollection idents = identCollection {formalParams = idents ++ formalParams identCollection}
+
+addLocalVars :: IdentCollection -> [Ident] -> IdentCollection
+addLocalVars identCollection idents = identCollection {localVars = idents ++ localVars identCollection}
+
+instance Transform ClassDecl where
+  analyze scope classDecl@(ClassDecl src modifiers ident typeParams mbRefType refTypes classBody) =
+    let newScope = addFields scope classDecl
+     in ClassDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          typeParams
+          mbRefType
+          refTypes
+          (analyze newScope classBody)
+  analyze scope recordDecl@(RecordDecl src modifiers ident typeParams recordFieldDecls refTypes classBody) =
+    let newScope = addFields scope recordDecl
+     in RecordDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          typeParams
+          recordFieldDecls
+          refTypes
+          (analyze newScope classBody)
+  analyze scope enumDecl@(EnumDecl src modifiers ident refTypes enumBody) =
+    let newScope = addFields scope enumDecl
+     in EnumDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          refTypes
+          (analyze newScope enumBody)
+
+instance Transform MemberDecl where
+  analyze scope (FieldDecl span modifiers type_ varDecls) = FieldDecl span (map (analyze scope) modifiers) type_ (map (analyze scope) varDecls)
+  analyze scope methodDecl@(MethodDecl span modifiers typeParams mbType ident formalParams exceptionTypes mbExp methodBody) =
+    MethodDecl
+      span
+      (map (analyze scope) modifiers)
+      typeParams
+      mbType
+      ident
+      (map (analyze scope) formalParams)
+      exceptionTypes
+      (fmap (analyze scope) mbExp)
+      (analyze (addFormalParams scope (methodParameters methodDecl)) methodBody)
+  analyze scope constructorDecl@(ConstructorDecl span modifiers typeParams ident formalParams exceptionTypes constructorBody) =
+    ConstructorDecl
+      span
+      (map (analyze scope) modifiers)
+      typeParams
+      ident
+      (map (analyze scope) formalParams)
+      exceptionTypes
+      (analyze (addFormalParams scope (methodParameters constructorDecl)) constructorBody)
+  analyze scope (MemberClassDecl classDecl) = MemberClassDecl (analyze scope classDecl)
+  analyze scope (MemberInterfaceDecl interfaceDecl) = MemberInterfaceDecl (analyze scope interfaceDecl)
+
+instance Transform Catch where
+  analyze scope catch@(Catch formalParam block) =
+    Catch
+      (analyze scope formalParam)
+      (analyze (addFormalParams scope (exceptionParameter catch)) block)
+
+instance Transform Block where
+  analyze scope block@(Block blockstmts) = Block (map (analyze (addLocalVars scope (localVarsInBlock block))) blockstmts)
```

Comment:
Nach dieser Regel wären alle Variablen, die in einem Block deklariert werden, im gesamten Block sichtbar. Das stimmt aber ja nicht. Bitte einen Testfall konstruieren, der ggf. diese Problem aufzeigt.